extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/BaseLeafFrameProcessor.java

```diff
@@ -130,8 +130,10 @@ protected abstract ReturnOrAwait<Unit> runWithInputChannel(
    * Helper intended to be used by subclasses. Applies {@link #segmentMapFn}, which applies broadcast joins
    * if applicable to this query.
    */
-  protected SegmentReference mapSegment(final Segment segment)
+  protected Optional<Segment> mapSegment(final Segment segment)
   {
-    return segmentMapFn.apply(ReferenceCountingSegment.wrapRootGenerationSegment(segment));
+    // we use wrapUnmanaged here because segment reference tracking and lifecycle management happens elsewhere, so all
+    // we need to be able to do here is apply a segment map function since we don't care about the provider
+    return segmentMapFn.apply(ReferenceCountedSegmentProvider.wrapUnmanaged(segment));
```

Comment:
Is this ok in situations where the `segmentMapFn` acquires its own references to things, such as joinables? (There might have been a problem with this even prior to this patch.)

Reply:
yea, the segment returned by this is registered with the closer (so additional references acquired will be released), but we do not need to close the `ReferenceCountedSegmentProvider` itself that we need to wrap the `Segment` we provided. This was previously a problem, we were not acquiring references on the joinables at all, but is solved by the mapping and acquiring being part of the same operation now